In [181]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

Visualizar la información de la base de datos.

In [182]:
df = pd.read_csv('telecom_users.csv')

In [184]:
df.head(5)

Unnamed: 0  customerID  gender  SeniorCitizen Partner Dependents  tenure  \
0        1869  7010-BRBUU    Male              0     Yes        Yes      72   
1        4528  9688-YGXVR  Female              0      No         No      44   
2        6344  9286-DOJGF  Female              1     Yes         No      38   
3        6739  6994-KERXL    Male              0      No         No       4   
4         432  2181-UAESM    Male              0      No         No       2   

  PhoneService MultipleLines InternetService  ...     DeviceProtection  \
0          Yes           Yes              No  ...  No internet service   
1          Yes            No     Fiber optic  ...                  Yes   
2          Yes           Yes     Fiber optic  ...                   No   
3          Yes            No             DSL  ...                   No   
4          Yes            No             DSL  ...                  Yes   

           TechSupport          StreamingTV      StreamingMovies  \
0  No internet service  No internet service  No internet service   
1                   No                  Yes                   No   
2                   No                   No                   No   
3                   No                   No                  Yes   
4                   No                   No                   No   

         Contract PaperlessBilling              PaymentMethod MonthlyCharges  \
0        Two year               No    Credit card (automatic)          24.10   
1  Month-to-month              Yes    Credit card (automatic)          88.15   
2  Month-to-month              Yes  Bank transfer (automatic)          74.95   
3  Month-to-month              Yes           Electronic check          55.90   
4  Month-to-month               No           Electronic check          53.45   

  TotalCharges  Churn  
0      1734.65     No  
1       3973.2     No  
2      2869.85    Yes  
3        238.5     No  
4        119.5     No  

[5 rows x 22 columns]

Renombrar las columnas

In [186]:
df.columns

Index(['Unnamed: 0', 'customerID', 'gender', 'SeniorCitizen', 'Partner',
       'Dependents', 'tenure', 'PhoneService', 'MultipleLines',
       'InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies', 'Contract',
       'PaperlessBilling', 'PaymentMethod', 'MonthlyCharges', 'TotalCharges',
       'Churn'],
      dtype='object')

In [187]:
df.columns = ['#', 'ID Cliente', 'Genero', 'Jubilado', 'En matrimonio',
       'Dependiente', 'Antiguedad(meses)', 'Servicio telefónico', 'Multiples lineas',
       'Servicio de Internet', 'Navegación segura', 'Copia de seguridad', 'Proteción de datos',
       'Soporte técnico', 'Servicio de Streaming', 'Cine por Streaming', 'Tipo de contrato',
       'Facturación electrónica', 'Método de pago', 'Pago mensual', 'Totalhistórico',
       'Abandono']

In [188]:
df.columns

Index(['#', 'ID Cliente', 'Genero', 'Jubilado', 'En matrimonio', 'Dependiente',
       'Antiguedad(meses)', 'Servicio telefónico', 'Multiples lineas',
       'Servicio de Internet', 'Navegación segura', 'Copia de seguridad',
       'Proteción de datos', 'Soporte técnico', 'Servicio de Streaming',
       'Cine por Streaming', 'Tipo de contrato', 'Facturación electrónica',
       'Método de pago', 'Pago mensual', 'Totalhistórico', 'Abandono'],
      dtype='object')

Tipo de dato de cada columna

In [190]:
df.dtypes

#                            int64
ID Cliente                  object
Genero                      object
Jubilado                     int64
En matrimonio               object
Dependiente                 object
Antiguedad(meses)            int64
Servicio telefónico         object
Multiples lineas            object
Servicio de Internet        object
Navegación segura           object
Copia de seguridad          object
Proteción de datos          object
Soporte técnico             object
Servicio de Streaming       object
Cine por Streaming          object
Tipo de contrato            object
Facturación electrónica     object
Método de pago              object
Pago mensual               float64
Totalhistórico              object
Abandono                    object
dtype: object

Eliminar filas con datos nulos

In [191]:
df.drop(df[df['Totalhistórico']==" "].index,inplace = True)

Eliminar columnas con datos inecesarios

In [194]:
df.drop(['#','ID Cliente'],axis=1,inplace=True)

In [196]:
df.head(4)

Genero  Jubilado En matrimonio Dependiente  Antiguedad(meses)  \
0    Male         0           Yes         Yes                 72   
1  Female         0            No          No                 44   
2  Female         1           Yes          No                 38   
3    Male         0            No          No                  4   

  Servicio telefónico Multiples lineas Servicio de Internet  \
0                 Yes              Yes                   No   
1                 Yes               No          Fiber optic   
2                 Yes              Yes          Fiber optic   
3                 Yes               No                  DSL   

     Navegación segura   Copia de seguridad   Proteción de datos  \
0  No internet service  No internet service  No internet service   
1                   No                  Yes                  Yes   
2                   No                   No                   No   
3                   No                   No                   No   

       Soporte técnico Servicio de Streaming   Cine por Streaming  \
0  No internet service   No internet service  No internet service   
1                   No                   Yes                   No   
2                   No                    No                   No   
3                   No                    No                  Yes   

  Tipo de contrato Facturación electrónica             Método de pago  \
0         Two year                      No    Credit card (automatic)   
1   Month-to-month                     Yes    Credit card (automatic)   
2   Month-to-month                     Yes  Bank transfer (automatic)   
3   Month-to-month                     Yes           Electronic check   

   Pago mensual Totalhistórico Abandono  
0         24.10        1734.65       No  
1         88.15         3973.2       No  
2         74.95        2869.85      Yes  
3         55.90          238.5       No

Convertir las columnas con datos que contienen palabras unicas a datos numericos

In [160]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [161]:
columnas = ['Genero', 'Jubilado', 'En matrimonio',
       'Dependiente', 'Servicio telefónico', 'Multiples lineas',
       'Servicio de Internet', 'Navegación segura', 'Copia de seguridad', 'Proteción de datos',
       'Soporte técnico', 'Servicio de Streaming', 'Cine por Streaming', 'Tipo de contrato',
       'Facturación electrónica','Método de pago','Abandono']

In [197]:
for i in columnas:
    df[i] = LabelEncoder().fit_transform(df[i])

In [198]:
df.head(4)

Genero  Jubilado  En matrimonio  Dependiente  Antiguedad(meses)  \
0       1         0              1            1                 72   
1       0         0              0            0                 44   
2       0         1              1            0                 38   
3       1         0              0            0                  4   

   Servicio telefónico  Multiples lineas  Servicio de Internet  \
0                    1                 2                     2   
1                    1                 0                     1   
2                    1                 2                     1   
3                    1                 0                     0   

   Navegación segura  Copia de seguridad  Proteción de datos  Soporte técnico  \
0                  1                   1                   1                1   
1                  0                   2                   2                0   
2                  0                   0                   0                0   
3                  0                   0                   0                0   

   Servicio de Streaming  Cine por Streaming  Tipo de contrato  \
0                      1                   1                 2   
1                      2                   0                 0   
2                      0                   0                 0   
3                      0                   2                 0   

   Facturación electrónica  Método de pago  Pago mensual Totalhistórico  \
0                        0               1         24.10        1734.65   
1                        1               1         88.15         3973.2   
2                        1               0         74.95        2869.85   
3                        1               2         55.90          238.5   

   Abandono  
0         0  
1         0  
2         1  
3         0

In [163]:
df.dtypes

Genero                       int32
Jubilado                     int64
En matrimonio                int32
Dependiente                  int32
Antiguedad(meses)            int64
Servicio telefónico          int32
Multiples lineas             int32
Servicio de Internet         int32
Navegación segura            int32
Copia de seguridad           int32
Proteción de datos           int32
Soporte técnico              int32
Servicio de Streaming        int32
Cine por Streaming           int32
Tipo de contrato             int32
Facturación electrónica      int32
Método de pago               int32
Pago mensual               float64
Totalhistórico              object
Abandono                     int32
dtype: object

In [164]:
df["Totalhistórico"] = pd.to_numeric(df["Totalhistórico"], downcast="float")

In [176]:
df.dtypes

Genero                       int32
Jubilado                     int64
En matrimonio                int32
Dependiente                  int32
Antiguedad(meses)            int64
Servicio telefónico          int32
Multiples lineas             int32
Servicio de Internet         int32
Navegación segura            int32
Copia de seguridad           int32
Proteción de datos           int32
Soporte técnico              int32
Servicio de Streaming        int32
Cine por Streaming           int32
Tipo de contrato             int32
Facturación electrónica      int32
Método de pago               int32
Pago mensual               float64
Totalhistórico             float32
Abandono                     int32
dtype: object

In [199]:
df

Genero  Jubilado  En matrimonio  Dependiente  Antiguedad(meses)  \
0          1         0              1            1                 72   
1          0         0              0            0                 44   
2          0         1              1            0                 38   
3          1         0              0            0                  4   
4          1         0              0            0                  2   
...      ...       ...            ...          ...                ...   
5981       1         0              1            0                  1   
5982       0         0              1            1                 23   
5983       1         0              1            1                 12   
5984       1         1              0            0                 12   
5985       1         0              0            0                 26   

      Servicio telefónico  Multiples lineas  Servicio de Internet  \
0                       1                 2                     2   
1                       1                 0                     1   
2                       1                 2                     1   
3                       1                 0                     0   
4                       1                 0                     0   
...                   ...               ...                   ...   
5981                    1                 0                     1   
5982                    1                 2                     0   
5983                    1                 0                     2   
5984                    1                 2                     1   
5985                    1                 0                     2   

      Navegación segura  Copia de seguridad  Proteción de datos  \
0                     1                   1                   1   
1                     0                   2                   2   
2                     0                   0                   0   
3                     0                   0                   0   
4                     2                   0                   2   
...                 ...                 ...                 ...   
5981                  2                   0                   0   
5982                  2                   2                   2   
5983                  1                   1                   1   
5984                  0                   0                   2   
5985                  1                   1                   1   

      Soporte técnico  Servicio de Streaming  Cine por Streaming  \
0                   1                      1                   1   
1                   0                      2                   0   
2                   0                      0                   0   
3                   0                      0                   2   
4                   0                      0                   0   
...               ...                    ...                 ...   
5981                0                      2                   2   
5982                2                      2                   2   
5983                1                      1                   1   
5984                0                      2                   2   
5985                1                      1                   1   

      Tipo de contrato  Facturación electrónica  Método de pago  Pago mensual  \
0                    2                        0               1         24.10   
1                    0                        1               1         88.15   
2                    0                        1               0         74.95   
3                    0                        1               2         55.90   
4                    0                        0               2         53.45   
...                ...                      ...             ...           ...   
5981                 0                        1               2         95.00   
5982                 2        

Entrenar el modelo de regresión logistica

In [166]:
y = df['Abandono'] #Columna objetivo

In [167]:
X = df.drop('Abandono', axis=1) #Caracteristicas

In [168]:
from sklearn.model_selection import train_test_split
#70% entrenamiento, 30% prueba, toma los datos de forma aleatoria
X_train, X_test, y_train, y_test = train_test_split(X,y,test_size=0.3,random_state=45)

In [169]:
from sklearn.linear_model import LogisticRegression

In [170]:
modelo = LogisticRegression()
modelo.fit(X_train, y_train)

C:\Users\Carlos\Anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:762: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [171]:
predicciones = modelo.predict(X_test)

In [172]:
predicciones #Valores con prediccion

array([0, 0, 1, ..., 0, 0, 0])

In [173]:
y_test #Valores reales

5586    0
4282    0
5489    1
2225    0
2055    1
       ..
1597    0
5378    1
1703    0
1124    0
1667    0
Name: Abandono, Length: 1793, dtype: int32

In [174]:
from sklearn.metrics import classification_report

In [175]:
print(classification_report(y_test,predicciones))

              precision    recall  f1-score   support

           0       0.85      0.88      0.86      1336
           1       0.61      0.53      0.57       457

    accuracy                           0.79      1793
   macro avg       0.73      0.71      0.72      1793
weighted avg       0.79      0.79      0.79      1793



precision de 88% para clientes actuales

precision de 61% para clientes que cancelaron el servicio